### PR TITLE
Minimal Version of Time Sync Cluster

### PR DESCRIPTION
**Pull request by ghost**
#### Problem
Currently the SDK lacks of support for the Time Sync Cluster. This PR tries to add a minimal version of that. 

#### Change overview
- Added correct XML Definition
- Added support for the chip-device-ctrl
- Added minimal impl. of Time Sync Cluster

#### Current Limitations
- Cannot support more then one node id (currently node id 0 as the spec suggests)
- No support for Time Zone, NTP Client or NTP Server.
- No support for Time Synchronization during Commissioning. 
- Time Synchronization is lost after a reboot of the device.

#### Testing
Tested with the all-clusters-app

#### Related
https://github.com/project-chip/connectedhomeip/issues/2871
( https://github.com/project-chip/connectedhomeip/issues/2872 )

### CODE REVIEW

File: examples/all-clusters-app/all-clusters-common/all-clusters-app.zap [hunk 1]
```diff
@@ -3365,6 +3365,172 @@
...
+              "mfgCode": null,
+              "side": "server",
+              "included": 1,
+              "storageOption": "RAM",
+              "singleton": 1,
+              "bounded": 0,
+              "defaultValue": "",
+              "reportable": 0,
+              "minInterval": 1,
+              "maxInterval": 65534,
```
Comment (cecille) [inline examples/all-clusters-app/all-clusters-common/all-clusters-app.zap:3424]: Apologies - I'm not super familiar with zap definitions so it's completely possible I'm misinterpreting the fields. Is this the max value for the field? If so, this should be much larger - this field is a uint64.  Ditto for the localtime field below.

File: examples/all-clusters-app/all-clusters-common/all-clusters-app.zap [hunk 2]
```diff
@@ -3365,6 +3365,172 @@
...
```
Comment (cecille) [inline examples/all-clusters-app/all-clusters-common/all-clusters-app.zap:3439]: If this is max value for the field, it should be either max of enum8 or values of 0-4 (enum values in the spec). Again, sorry, I'm not familiar with the zap conventions here, so I'm not totally clear on what this represents or what the convention for enums is. 

Ditto with the time source below - it's an enum8 with values from 0 to 16.

File: src/app/clusters/time-sync-server/time-sync-server.cpp [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+}
+} // anonymous namespace
+
+bool emberAfTimeSyncClusterSetUtcTimeCallback(chip::EndpointId endpoint, chip::app::CommandHandler * commandObj, uint64_t UtcTime,
+                                              uint8_t Granularity, uint8_t TimeSource)
+{
+    EmberAfStatus status = EMBER_ZCL_STATUS_SUCCESS;
+    uint8_t granularity  = EMBER_ZCL_TIME_SYNC_GRANULARITY_NO_TIME_GRANULARITY;
+
+    if (EMBER_ZCL_STATUS_SUCCESS == TimeSync::Attributes::SetGranularity(endpoint, granularity) &&
```
Comment (cecille) [inline src/app/clusters/time-sync-server/time-sync-server.cpp:99]: Is this supposed to be GetGranularity? I might be misunderstanding. What does this check do? It looks like it unconditionally sets our current granularity to NO_TIME_GRANULARITY.

File: src/app/clusters/time-sync-server/time-sync-server.cpp [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    uint64_t GetUtcTime(void) { return (GetLocalTime() - mTimeStamp) + mTimeBase; }
+    uint64_t GetLocalTime(void) { return System::Clock::GetMonotonicMicroseconds(); }
+    void SetTimeBase(uint64_t timeBase)
+    {
+        mTimeBase  = timeBase;
+        mTimeStamp = GetLocalTime();
+    }
+
+private:
+    uint64_t mTimeBase  = 0; // TODO store TimeBase in Flash ?
```
Comment (cecille) [inline src/app/clusters/time-sync-server/time-sync-server.cpp:44]: What are the time stamp and time base?

File: src/app/clusters/time-sync-server/time-sync-server.cpp [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+#include <system/SystemClock.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+
+struct TimeSyncTimeSource
+{
+    uint64_t GetUtcTime(void) { return (GetLocalTime() - mTimeStamp) + mTimeBase; }
+    uint64_t GetLocalTime(void) { return System::Clock::GetMonotonicMicroseconds(); }
```
Comment (cecille) [inline src/app/clusters/time-sync-server/time-sync-server.cpp:36]: Local time is meant to be a DST/TZ adjusted absolute time, not the time since boot.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

**Comment by jmeg-sfy** [comment]
not stalled

**Pull request closed without merging**